aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright © Aptos Foundation
+
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    natives::{
+        cryptography::algebra::gas::GasParameters,
+        helpers::{make_safe_native, SafeNativeContext, SafeNativeError, SafeNativeResult},
+    },
+    safely_pop_arg,
+};
+use aptos_types::on_chain_config::{FeatureFlag, Features, TimedFeatures};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use better_any::{Tid, TidAble};
+use move_core_types::language_storage::TypeTag;
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    values::{Value, VectorRef},
+};
+use smallvec::{smallvec, SmallVec};
+use std::{any::Any, collections::VecDeque, hash::Hash, ops::Add, rc::Rc, sync::Arc};
+
+pub mod gas;
+
+/// Equivalent to `std::error::not_implemented(0)` in Move.
+const MOVE_ABORT_CODE_NOT_IMPLEMENTED: u64 = 0x0C0000;
+
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub enum Structure {
+    BLS12381Fq12,
+    BLS12381G1Affine,
+    BLS12381G2Affine,
+    BLS12381Gt,
+    BLS12381Fr,
+}
+
+impl Structure {
+    pub fn from_type_tag(type_tag: &TypeTag) -> Option<Structure> {
+        match type_tag.to_string().as_str() {
+            "0x1::algebra_bls12381::Fr" => Some(Structure::BLS12381Fr),
+            "0x1::algebra_bls12381::Fq12" => Some(Structure::BLS12381Fq12),
+            "0x1::algebra_bls12381::G1Affine" => Some(Structure::BLS12381G1Affine),
+            "0x1::algebra_bls12381::G2Affine" => Some(Structure::BLS12381G2Affine),
+            "0x1::algebra_bls12381::Gt" => Some(Structure::BLS12381Gt),
```

Comment:
@vgao1996 @wrwg is it a good practice to rely on the full path of a type parameter in native backend?

Reply:
This one's interesting as it's the first time we're trying to implement natives that are specialized over the parameter types. 

Well, when you call `fn all_natives`, you pass in a framework_addr, so conceptually it's better to check again this address rather than the hardcoded "0x1".

Reply:
This would technically allow the natives to be used under a different address, but I'm not sure how useful that is though.

Reply:
Our `Table` natives also rely on the type info fetched via `context.type_to_type_tag`. So if the format changes, all of the tables in storage will stop working.

Similarly, in Move, we rely on the type name (also fetched via  `context.type_to_type_tag`) in `any.move` and in `property_map.move`.

I don't think we are breaking new ground here in terms of relying on type names, although we are breaking new ground in terms of specializing a native based on a template arg.

But I do agree with Victor that it would be better not to rely on a hardcoded address... We can pass in the address to all the natives and then adjust the `from_type_tag(type_tag: &TypeTag)` function to take it as a parameter.

Reply:
This will have to be done in a separate PR because it will touch `make_safe_native` and all other natives.